config/v1/0000_10_config-operator_01_infrastructure-CustomNoUpgrade.crd.yaml

```diff
@@ -507,6 +507,13 @@ spec:
                                 type: string
                             type: object
                           type: array
+                        userProvidedDNS:
+                          default: Disabled
+                          description: 'UserProvidedDNS indicates if the customer is providing their own DNS solution in place of the default provided by the underlying platform. Its value is set by the Installer and can have 2 possible values: "Enabled" and "Disabled" with "Disabled" being the default.'
```

Comment:
>
>`I think this is an important distinction, because it is completely possible (and I think not uncommon) for >users to provide their own custom local DNS servers.`
>
I can see where this comment is coming from. Firstly, let us avoid using "custom" in many different contexts . That makes design around this feature really confusing.
The alternative that is being suggested above : "users provide their own local DNS servers. This case should be qualified with : Does the user configure this local DNS solution completely by themselves OR do they want OpenShift to configure it?"

The goal of this specific feature is to allow users to bring their own DNS solutions that OpenShift is *not* allowed to program. This is the distinction that is being missed in the comments above. 

So, let us look into the states this config can have:
1. Do you agree that programming the DNS solution provided by the cloud is the Default behavior of OpenShift cluster install on AWS, Azure or GCP? If yes, then whatever this config item is called, its default value should represent this state. Proposed: "CloudDefault/PlatformDefault"

2. The other end of the spectrum would be the case where the user provides a DNS before cluster install and no intervention is needed by OpenShift. Proposed: "UserProvided/UserProvisioned" (Example: UPI) **[In the case of UPI, who sets this state within the Infrastructure CR. The user? Is there any component in OpenShift that will make a decision based on this state?]**

3. Now this feature. User provides their own DNS solution (post-install because they won't have LB IPs before install) that OpenShift cannot program. So, OpenShift installs itself by starting its own in-cluster DNS and the user provides a DNS post-install and they co-exist. Proposed: "UserAndClusterProvided".

Reply:
> The goal of this specific feature is to allow users to bring their own DNS solutions that OpenShift is not allowed to program. This is the distinction that is being missed in the comments above.

OpenShift would be using CoreDNS to program DNS (on the cluster network). I don't think the distinction is missed; the entire point of my comment is that the current API design does not accurately describe this state (where OpenShift doesn't program DNS at all). I also acknowledge I'm not sure we need this state, but I think it should be considered.

> Do you agree that programming the DNS solution provided by the cloud is the Default behavior of OpenShift cluster install on AWS, Azure or GCP? If yes, then whatever this config item is called, its default value should represent this state. Proposed: "CloudDefault/PlatformDefault"

For IPI: yes; but for UPI: no. But a more generic value could accurately capture this state. Perhaps it helps to tie this value to installation `InstallProvided` or `InstallDefault` or some such? 
 
> The other end of the spectrum would be the case where the user provides a DNS before cluster install and no intervention is needed by OpenShift. Proposed: "UserProvided/UserProvisioned" (Example: UPI) [In the case of UPI, who sets this state within the Infrastructure CR. The user? Is there any component in OpenShift that will make a decision based on this state?]

IIUC there is no difference in behavior within the cluster for this value and the previous value (in both cases coreDNS remains off). This is true for the changes I proposed as well. 

As an aside, IMHO we're overcomplicating this by trying to include whether public DNS was user- or installer-provisioned in the API, but we have no motivation to do that. If we restrict the scope of the API field to whether on-cluster DNS is enabled or not, a default value of `Disabled` would always be true for UPI or IPI, and `Enabled` would allow us to enable coreDNS.

Another alternative would be to trigger CoreDNS on/off based on whether the LB IP/DNS name values are populated.